File: Q3_Shadow_DOM/automation/scripts/shadow_pathfinder.js
/**
 * Recursive Shadow DOM Pathfinder
 * Traverses open ShadowRoots to any depth to locate a target element using stable semantic properties.
 * 
 * Rules:
 * - NO element IDs
 * - NO absolute XPath
 * - NO CSS host classes/tags
 * - NO plain text matching
 * - Traverses OPEN shadow roots recursively
 */

(function () {
  window.findTargetInShadowDOM = function (targetSemantics) {
    const semantics = targetSemantics || {
      role: 'button',
      ariaLabel: 'Authorize Ledger Funds',
      qaState: 'unlocked-token'
    };

    let foundTarget = null;
    let maxDepthReached = 0;

    function searchRecursive(node, currentDepth) {
      if (!node) return false;

      if (currentDepth > maxDepthReached) {
        maxDepthReached = currentDepth;
      }

      // Check if current node matches target semantic criteria
      if (node.nodeType === 1) { // Node.ELEMENT_NODE
        const role = node.getAttribute('role');
        const ariaLabel = node.getAttribute('aria-label');
        const qaState = node.getAttribute('data-qa-state');

        const matchesRole = !semantics.role || role === semantics.role;
        const matchesAria = !semantics.ariaLabel || ariaLabel === semantics.ariaLabel;
        const matchesQaState = !semantics.qaState || qaState === semantics.qaState;

        if (matchesRole && matchesAria && matchesQaState) {
          foundTarget = {
            element: node,
            role: role,
            ariaLabel: ariaLabel,
            qaState: qaState,
            tagName: node.tagName.toLowerCase(),
            shadowDepth: currentDepth
          };
          return true; // Target found
        }
      }

      // 1. If node hosts an OPEN shadowRoot, traverse into shadowRoot first
      if (node.shadowRoot) {
        if (searchRecursive(node.shadowRoot, currentDepth + 1)) {
          return true;
        }
      }

      // 2. Traverse element children
      const children = Array.from(node.children || node.childNodes || []).filter(n => n.nodeType === 1);
      for (const child of children) {
        if (searchRecursive(child, currentDepth)) {
          return true;
        }
      }

      return false;
    }

    const startRoot = document.body || document.documentElement;
    const success = searchRecursive(startRoot, 0);

    if (success && foundTarget) {
      return {
        found: true,
        role: foundTarget.role,
        ariaLabel: foundTarget.ariaLabel,
        qaState: foundTarget.qaState,
        tagName: foundTarget.tagName,
        shadowDepth: foundTarget.shadowDepth,
        maxDepthReached: maxDepthReached
      };
    }

    return {
      found: false,
      shadowDepth: 0,
      maxDepthReached: maxDepthReached
    };
  };
})();

File: Q3_Shadow_DOM/automation/shadow_pathfinder.py
import os
from playwright.async_api import Page

class ShadowPathfinder:
    """
    Python wrapper executing the recursive Shadow DOM pathfinder script in Playwright.
    """
    def __init__(self, page: Page):
        self.page = page
        self.script_path = os.path.join(
            os.path.dirname(__file__), "scripts", "shadow_pathfinder.js"
        )

    async def inject_script(self):
        """Injects shadow_pathfinder.js into the browser runtime."""
        with open(self.script_path, "r", encoding="utf-8") as f:
            script_content = f.read()
        await self.page.evaluate(script_content)

    async def find_target(self, role: str = "button", aria_label: str = "Authorize Ledger Funds", qa_state: str = "unlocked-token") -> dict:
        """
        Executes recursive Shadow DOM search targeting element with stable semantic properties.
        """
        await self.inject_script()
        target_semantics = {
            "role": role,
            "ariaLabel": aria_label,
            "qaState": qa_state
        }
        result = await self.page.evaluate(
            "semantics => window.findTargetInShadowDOM(semantics)",
            target_semantics
        )
        return result
